corehq/motech/fhir/repeaters.py

```diff
@@ -0,0 +1,162 @@
+)
+from corehq.privileges import DATA_FORWARDING
+from corehq.toggles import FHIR_INTEGRATION
+
+from .const import FHIR_VERSION_4_0_1, XMLNS_FHIR
+from .models import FHIRResourceType
+from .repeater_helpers import get_fhir_bundle, register_patients, send_bundle
+
+
+class FHIRRepeater(CaseRepeater):
+    class Meta:
+        app_label = 'repeaters'
+
+    friendly_name = _('Forward Cases to a FHIR API')
+    payload_generator_classes = (FormDictPayloadGenerator,)
+    include_app_id_param = False
+    _has_config = False
+
+    fhir_version = StringProperty(default=FHIR_VERSION_4_0_1)
+
+    @memoized
+    def payload_doc(self, repeat_record):
+        return FormAccessors(repeat_record.domain).get_form(repeat_record.payload_id)
+
+    @property
+    def form_class_name(self):
+        # The class name used to determine which edit form to use
+        return self.__class__.__name__
+
+    @classmethod
+    def available_for_domain(cls, domain):
+        return (domain_has_privilege(domain, DATA_FORWARDING)
+                and FHIR_INTEGRATION.enabled(domain))
+
+    def allowed_to_forward(self, payload):
+        # When we update a case's external_id to their ID on a remote
+        # FHIR service, the form is submitted with XMLNS_FHIR. This
+        # check makes sure that we don't send the update back to FHIR.
+        return payload.xmlns != XMLNS_FHIR
+
+    def send_request(self, repeat_record, payload):
+        """
+        Generates FHIR resources from ``payload``, and sends them as a
+        FHIR transaction bundle. If there are patients that need to be
+        registered, that is done first.
+
+        Returns an HTTP response-like object. If the payload has nothing
+        to send, returns True.
+        """
+        response = True
+        requests = self.connection_settings.get_requests(repeat_record.payload_id)
+        case_trigger_infos = self.get_case_trigger_infos(
+            payload,
+            self.fhir_version,
+        )
+        try:
+            register_patients(requests, case_trigger_infos, self.fhir_version)
+            bundle = get_fhir_bundle(case_trigger_infos, self.fhir_version)
+            if bundle:
+                response = send_bundle(requests, bundle)
+        except Exception as err:
+            requests.notify_exception(str(err))
+            return RepeaterResponse(400, 'Bad Request', pformat_json(str(err)))
+        return response
+
+    def get_case_trigger_infos(self, form_json, fhir_version):
+        form_question_values = get_form_question_values(form_json)
+        case_blocks = extract_case_blocks(form_json)
+        cases_by_id = _get_cases_by_id(self.domain, case_blocks)
+        resource_types_by_case_type = _get_resource_types_by_case_type(
+            self.domain,
+            fhir_version,
+            cases_by_id.values(),
+        )
+
+        case_trigger_infos = []
+        for case_block in case_blocks:
+            case = cases_by_id[case_block['@case_id']]
```

Comment:
This would fail if the database had become inconsistent -- form processing was failing to create cases, or a case had been hard-deleted, but its forms remained or it still existed on a device. Since this code is triggered on a new form submission, it would be less likely to encounter old data.

Is this the right place to be checking for those possibilities? I've changed this to raise `CaseNotFound`, but I'm not certain this is the right approach. 741e5569481 

I'm intentionally keeping this outside the `try ... except` block in `FHIRRepeater.send_request()` that notifies third-party administrators of errors.